url/src/host.rs

```diff
@@ -137,17 +154,39 @@ impl Host<String> {
             Err(ParseError::InvalidDomainCharacter)
         } else {
             Ok(Host::Domain(
-                utf8_percent_encode(input, CONTROLS).to_string(),
+                match utf8_percent_encode(&input, CONTROLS).into() {
+                    Cow::Owned(v) => Cow::Owned(v),
+                    // if we're borrowing, then we can return the original Cow
+                    Cow::Borrowed(_) => input,
+                },
             ))
         }
     }
 
-    /// convert domain with idna
-    fn domain_to_ascii(domain: &[u8]) -> Result<Cow<'_, str>, ParseError> {
-        idna::domain_to_ascii_cow(domain, idna::AsciiDenyList::URL).map_err(Into::into)
+    pub(crate) fn into_owned(self) -> Host<String> {
+        match self {
+            Host::Domain(s) => Host::Domain(s.into_owned()),
+            Host::Ipv4(ip) => Host::Ipv4(ip),
+            Host::Ipv6(ip) => Host::Ipv6(ip),
+        }
     }
 }
 
+/// convert domain with idna
+fn domain_to_ascii(domain: Cow<'_, [u8]>) -> Result<Cow<'_, str>, ParseError> {
+    let value = idna::domain_to_ascii_cow(&domain, idna::AsciiDenyList::URL)?;
+    // TODO: would be better to move this into the idna crate
+    Ok(match value {
+        Cow::Owned(value) => Cow::Owned(value),
+        // SAFETY: If borrowed, then the original string is ascii and we can return the
+        // original Cow in order to save an allocation
```

Comment:
My reading of the code suggests this is correct right now, but:

There is no guarantee that `idna::domain_to_ascii_cow` could not return a `Cow::Borrowed(x)` where `x.len() < domain.len()` in the future (where this code would break or possibly be unsafe). To avoid that refactoring hazard, we should hoist this up into `idna`.

Reply:
Hoisted. Not to sure about the new function name.